components/collector/src/base_collectors/collector.py

```
@@ -43,41 +40,37 @@ def record_health() -> None:
+        # is closed. But due to the architecture of the collector (first collect all the responses, then parse them)
+        # the responses are kept around relatively long, and hence the connections too. To prevent time-outs while
+        # aiohttp waits for connections to become available we don't limit the connection pool size.
+        async with aiohttp.ClientSession(
+            connector=aiohttp.TCPConnector(limit=0, ssl=False),
+            raise_for_status=True,
+            timeout=aiohttp.ClientTimeout(total=120),
+            trust_env=True,
+        ) as session:
+            while True:
+                self.__log_tasks("waking up")
+                self.record_health()
+                await self.collect_metrics(session)
+                self.__log_tasks(f"going to sleep for {config.MAX_SLEEP_DURATION} seconds", config.MEASUREMENT_LIMIT)
+                await asyncio.sleep(config.MAX_SLEEP_DURATION)
 
     async def collect_metrics(self, session: aiohttp.ClientSession) -> None:
         """Collect measurements for metrics, prioritizing edited metrics."""
         reports = get_reports(self.database)
         metrics = get_metrics_from_reports(reports)
         next_fetch = datetime.now(tz=UTC) + timedelta(seconds=config.MEASUREMENT_FREQUENCY)
-        tasks: list[Coroutine] = []
+        created_tasks = 0
         for metric_uuid, metric in self.__sorted_by_edit_status(cast(JSONDict, metrics)):
-            if len(tasks) >= config.MEASUREMENT_LIMIT:
-                break
             if self.__should_collect(metric_uuid, metric):
-                tasks.append(self.collect_metric(session, metric_uuid, metric, next_fetch))
-        await asyncio.gather(*tasks)
+                task = asyncio.create_task(self.collect_metric(session, metric_uuid, metric, next_fetch))
+                self.tasks.add(task)
+                created_tasks += 1
+                task.add_done_callback(self.tasks.discard)
```

Does this not also discard errors and stack traces?

---

No, it just calls `self.tasks.discard(task)` when the task is done.